### ISSUE TITLE
Rewrite Ken's/Rolfs American Options pricer in Futhark

### ISSUE THREAD

**Issue by Athas** [HIPERFIT/futhark#86]
@kfl made a Haskell and SML port of some of Rolfs R-code: https://github.com/kfl/american-options

It would be interesting to rewrite this in Futhark.

**Comment by kfl** [comment]
Yay, finally after three years somebody listens :fireworks:  :wink:

**Comment by Athas** [comment]
It will need some more work to make it efficient on a GPU before.  Maybe Ken has some ideas?

**Comment by Athas** [comment]
How on Earth did the Accelerate version ever manage to get efficient GPU code out of this with delayed arrays?

**Comment by Athas** [comment]
Some performance numbers for the `futhark-c`-code compared to the SML implementation:

```
$ time ./AmrPutVec 
AmrPut.binom(1) = 6.74543295136
AmrPut.binom(8) = 13.9456888384
AmrPut.binom(16) = 16.2225913859
AmrPut.binom(30) = 17.6537059071
AmrPut.binom(64) = 18.4299315651

real	0m2.773s
user	0m2.764s
sys	0m0.004s
```

```
$ time (for x in 1 8 16 30 64; do echo $x | ./american_option-cpu; done)
6.745433
13.945689
16.222591
17.653706
18.429932

real	0m1.308s
user	0m0.936s
sys	0m0.356s
```

The careful reader will note some discrepancies here.  First, I ditched the expiry=128 case, because Futhark still does very little memory optimisation and exactly no free()ing.  This hurts us very badly in the main loop, as Futhark (stupidly) creates an allocation for every iteration.  I tried to run this on the GPU machines, as they have much more memory than my laptop and thus should finish without swapping, but MLton is apparently not in the OpenSUSE package system, and it's too late in the evening to install an SML compiler manually.

Second, my Futhark timing is pretty lazy and measures the startup time several times.

When someone (probably future me) gets around to actually doing something with Futhark's fairly powerful internal memory representations, we can probably make these numbers much better.
